### ISSUE TITLE
Entering very small manual bolus crashes Loop on Medtronic

### ISSUE THREAD

**Issue by elnjensen** [LoopKit/Loop#1637]
**Describe the bug**
If the user enters a very small bolus (small enough to round to zero in `roundToSupportedBolusVolume`) and attempts to deliver it, the app will crash with an assertion failure at [this line](https://github.com/ps2/rileylink_ios/blob/16bad5a116c7f86526e21484c2db5789d62983d6/MinimedKit/PumpManager/MinimedPumpManager.swift#L1173). 

**To Reproduce**
1.  Click on the manual bolus icon. 
2. Enter 0.001 units. 
3. Press Done. 
4. Press Deliver. 

**Expected behavior**
App should not crash - perhaps an error/warning message listing the smallest deliverable volume for the configured pump, or simply stating that the bolus is too small to deliver. 

**Loop Version**
 - Latest dev as of 2022-03-13, 10:00 PM

**Pump**
 - Manufacturer: Medtronic
 - Model: 522
 
Fundamentally this is the user's fault of course, but could be handled more gracefully, as the user may not know what the smallest deliverable volume is.

**Comment by ps2** [comment]
I think maybe we should amend the roundToSupportedBolusVolume protocol to allow a return of nil if the value cannot be supported by the pump. Loop probably shouldn't even be trying to bolus if the bolus amount would fail.  Would you be up for that kind of change? I can provide guidance if needed.

**Comment by elnjensen** [comment]
OK, finally getting back to this.  Some questions, to try to understand what you're envisioning here. 

1.  In my original case of the user entering a small bolus that rounds to zero, should the protocol return nil rather than zero? 
2. If so, should there be a notification to the user, or just take no action silently? 
3. Should all cases of rounding to zero return nil instead?  It seems like that would overcomplicate things - e.g. if an auto-bolus recommendation would round to zero, that's fine.  The code should just note that and move on.  So presumably we would need to distinguish between how autoboluses and manual boluses are handled.  That would mean an extra argument to the `roundToSupportedBolusVolume` functions, which right now just take units as the argument. 
4. Similarly, those functions are used in the progress indicators, where again a 0 result would be fine but nil might be problematic.  
5. Are there other cases (besides being below the minimum increment) that result in an undeliverable dose after rounding?  It looks like manual bolus requests above the configured max bolus are already caught before this step - are there other cases I'm not thinking of? 

OK, those are my questions - happy to work on this, but I guess the big picture of all of the above is that it seems like there are cases where rounding to zero is useful and appropriate, and cases where it means we can't take action.  So I'm trying to understand which is which, and where that decision should get made.

**Comment by ps2** [comment]
Yes those are the main questions, and they're mostly the responsibility of the higher level code. In each case, the calling code should probably know what to do if it turns out that the "rounded" value is not deliverable. I wouldn't change the behavior of `roundToSupportedBolusVolume` based on automatic vs manual bolusing; the code in each of those two places would just deal with a returned nil differently if needed. Yes, I would return nil for all volumes that would round to zero, as zero is a non-sensical bolus amount. The other option is to remove that method from the pump managers, and have rounding happen in the higher level code, using the pumpmanager provided supported bolus volumes. I'm not sure why we have it down there anyways; might have been added before we were exposing the list of supported bolus volumes?

**Comment by ps2** [comment]
Sorry for putting this discussion onto the issue report; probably should've put it on the PR instead. Maybe we can continue there.
